fsspec/parquet.py

```diff
@@ -220,29 +283,81 @@ def _get_parquet_byte_ranges(
         for i, path in enumerate(paths):
             result[path] = {(footer_starts[i], footer_ends[i]): footer_samples[i]}
 
-    # Use cat_ranges to gather the data byte_ranges
-    for i, data in enumerate(fs.cat_ranges(data_paths, data_starts, data_ends)):
-        if data_ends[i] > data_starts[i]:
-            result[data_paths[i]][(data_starts[i], data_ends[i])] = data
+    # Transfer the data byte-ranges into local memory
+    _transfer_ranges(fs, result, data_paths, data_starts, data_ends)
 
     # Add b"PAR1" to header if necessary
     if add_header_magic:
-        for i, path in enumerate(paths):
-            add_magic = True
-            for k in result[path].keys():
-                if k[0] == 0 and k[1] >= 4:
-                    add_magic = False
-                    break
-            if add_magic:
-                result[path][(0, 4)] = b"PAR1"
-
-    # Add b"" for reads beyond end of file
-    for i, path in enumerate(paths):
-        result[path][(file_sizes[i], 2 * file_sizes[i])] = b""
+        _add_header_magic(result)
 
     return result
 
 
+def _get_parquet_byte_ranges_from_metadata(
+    metadata,
+    fs,
+    columns=None,
+    row_groups=None,
+    max_gap=64_000,
+    max_block=256_000_000,
+    engine="auto",
+):
+    """Simplified version of `_get_parquet_byte_ranges` for
+    the case that an engine-specific `metadata` object is
+    provided, and the remote footer metadata does not need to
+    be transfered before calculating the required byte ranges.
+    """
+
+    # Set engine if necessary
+    if isinstance(engine, str):
+        engine = _set_engine(engine)
```

Comment:
Engine is not used in multiple places? Is it possible to use one engine for the metadata and another for the data read?

Reply:
I removed this particular engine check, but I agree that there is further work to be done to properly check/infer that the correct engine is used with user-specified metadata.